### ISSUE TITLE
logerror message："SQLite error 14: unable to open database file”

### ISSUE THREAD

**Issue by slzButton** [msintuneappsdk/ms-intune-app-sdk-ios#197]
**Describe the bug:**
1.由于我们的app使用了voip的推送，并且需要在收到voip的推送的时候访问我们本地的数据文件进行解密，但是我们发现intune貌似阻拦了我们的操作，导致我们的sqlite报错，提示我们："SQLite error 14: unable to open database file”,附件报错信息
2.当我们的app绑定过intune后，并且在正常使用过程中，我们把app kill后，开启飞行模式，使网略链接断开，接着我们在无网络情况下进入app，会出现一个alert提示我们：离线宽限期过期。

1. Since our app uses voip push and needs to access our local data file for decryption when receiving voip push, but we found that Intune seems to block our operation, so our SQLite reported an error and said: "SQLite Error 14: Unable to open database file ", error message is reported in the attachment
2. When our app is bound to InTune and in normal use, we kill the app and turn on the flight mode to disconnect the Internet link. Then when we enter the app without network, an alert will appear to remind us that the offline grace period has expired.

**To Reproduce**
1.Configure some intune policies
2.Launch the intune app and complete enrollment with intune
3.After restart prompt, launch the app again and finish any policy prompts
4.Move the app to background

**Screenshots and logs:**
<img width="1315" alt="报错信息" src="https://user-images.githubusercontent.com/13009437/131458321-355c82b3-685d-47d5-9a50-1381a738516b.png">
[com.hillhouse.gyapp 2021-08-19--08-45-18-543.log](https://github.com/msintuneappsdk/ms-intune-app-sdk-ios/files/7081911/com.hillhouse.gyapp.2021-08-19--08-45-18-543.log)

**Comment by abasore** [comment]
Hi @slzButton are the two points you mentioned two separate issues? For the first one, do you have the 'MaxFileProtectionLevel' level set in the IntuneMAMSettings plist? Without setting that value, it defaults to the highest protection level (NSFileProtectionComplete) where protected files can only be accessed while the device is unlocked. Ten seconds after the device is locked, the app will lose access to protected files. The docs here describe this in more detail: https://docs.microsoft.com/en-us/mem/intune/developer/app-sdk-ios#configure-settings-for-the-intune-app-sdk

**Comment by abasore** [comment]
As for the second issue, that may be expected behavior. The docs [here](https://docs.microsoft.com/en-us/mem/intune/apps/app-protection-policy-settings-ios#conditional-launch) have info on the offline grace period policy.

**Comment by slzButton** [comment]
> 嗨@slzButton你提到的两点是两个不同的问题吗？对于第一个，您是否在 IntuneMAMSettings plist 中设置了“MaxFileProtectionLevel”级别？如果不设置该值，则默认为最高保护级别 (NSFileProtectionComplete)，只有在设备解锁时才能访问受保护的文件。设备锁定十秒后，该应用程序将无法访问受保护的文件。这里的文档更详细地描述了这一点：[https](https://docs.microsoft.com/en-us/mem/intune/developer/app-sdk-ios#configure-settings-for-the-intune-app-sdk) : [//docs.microsoft.com/en-us/mem/intune/developer/app-sdk-ios#configure-settings-for-the-intune-app-sdk](https://docs.microsoft.com/en-us/mem/intune/developer/app-sdk-ios#configure-settings-for-the-intune-app-sdk)

谢谢您的回复，我们按照您的回复设置了MaxFileProtectionLevel，并设置为NSFileProtectionNone，应该解决了我们的第一个问题。
Thank you for your reply. We set MaxFileProtectionLevel according to your reply and set it to NSFileProtectionNone, which should solve our first problem。

**Comment by slzButton** [comment]
> 至于第二个问题，这可能是预期的行为。[此处](https://docs.microsoft.com/en-us/mem/intune/apps/app-protection-policy-settings-ios#conditional-launch)的文档包含有关离线宽限期政策的信息。

![attachment-20210831-150230_01](https://user-images.githubusercontent.com/13009437/131654965-7b733b58-dbd3-4322-b509-1c9c368dd8f4.png)
关于第二个问题，我们设置了Offline grace period，Block access (minutes) ：720 minutes (12 hours)，Wipe data (days)：90 days，可是当我们的应用绑定了InTune并且正常使用时，我们杀死应用并打开飞行模式断开互联网链接。然后当我们进入没有网络的应用程序时，会出现一个警报，提醒我们离线宽限期已过。我感觉我们并没有超过离线宽限期720 minutes (12 hours)。

Block Access (minutes) : 720 minutes (12 hours). Wipe data (days) : 90 days, but when our app was tied to InTune and working, we killed the app and turned on airplane mode to disconnect from the Internet. Then when we enter an application without a network, an alert appears reminding us that the offline grace period has expired. I feel like we haven't exceeded the offline grace period 720 minutes (12 hours).

**Comment by abasore** [comment]
@slzButton Yes, you are correct grace period shouldn't show up before the 720 hours based on your policy. Just to check, is the device clock being set ahead at all during testing? Setting the device date/time ahead will trigger the grace period alert.

**Comment by slzButton** [comment]
> @slzButton Yes, you are correct grace period shouldn't show up before the 720 hours based on your policy. Just to check, is the device clock being set ahead at all during testing? Setting the device date/time ahead will trigger the grace period alert.

是的我们却是修改了设备的日期，但是当我们触发宽限期到期后后，再次将设备时间改回来，也无法进入app了，这是已有bug么？大概什么时候会修复呢？

Yes, we changed the date of the device, but when we changed the time of the device again after the grace period expired, we could not enter the APP. Is there a bug? When will it be repaired?

**Comment by abasore** [comment]
@slzButton Changing the device date forward and then backward may result in unexpected behavior. Have you tested without setting the device time at all? For example, let the grace period lapse by letting the device/app be inactive for the full grace period time? Then attempting to access the app?

**Comment by slzButton** [comment]
> @slzButton Changing the device date forward and then backward may result in unexpected behavior. Have you tested without setting the device time at all? For example, let the grace period lapse by letting the device/app be inactive for the full grace period time? Then attempting to access the app?

有没有什么办法能让我们出现了在设置了时间导致离线宽限期出现提示的时候，让它恢复呢？重启手机？重新设置时间？重装app？

Is there any way to restore the offline grace period when it is prompted by setting the time? Reboot your phone? Reset the time? Reinstall the app?

**Comment by slzButton** [comment]
我们今天还出现了一个新的问题，当我们在app内用相机拍照后，并对照片进行了编辑操作，然后我们将照片通过我们pickerviewcontroller选择后发送给另一个用户的时候提示我们操作权限有问题

We also had a new problem today. When we took a photo with the camera in the app and edited the photo, we were reminded that there was a problem with the operation permission when we sent the photo to another user after selecting it by our PickerViewController

![image](https://user-images.githubusercontent.com/13009437/133212417-43d841d5-dbd2-4db1-b2ae-f38a15e8f6d8.png)

**Comment by abasore** [comment]
@slzButton For the offline grace period issue, restarting the phone should hard reset the device time. This will reset the check for the offline grace period. Are you seeing any issue with the expected behavior when the device time is not set ahead (as in letting the grace period lapse naturally)?

**Comment by abasore** [comment]
@slzButton for the camera issue, could you open a new support item? I think that would help keep the discussions organization.

**Comment by slzButton** [comment]
> @slzButton For the offline grace period issue, restarting the phone should hard reset the device time. This will reset the check for the offline grace period. Are you seeing any issue with the expected behavior when the device time is not set ahead (as in letting the grace period lapse naturally)?

正常使用的时候我们不太担心这种情况出现，我现在遇到的情况如下：

https://user-images.githubusercontent.com/13009437/133561420-8f826595-0e38-4220-beb5-26ad380b390c.mp4

我们无法在不联网状态下让用户正常进入app
我们是否有什么办法能够设置取消离线宽限期，或者设置不去校验离线宽限期

In normal use, we don't worry too much about this situation. The situation I encountered is as follows:

https://user-images.githubusercontent.com/13009437/133561420-8f826595-0e38-4220-beb5-26ad380b390c.mp4

We cannot allow users to access the APP normally without being connected to the Internet
Is there any way to cancel the offline grace period, or not to validate the offline grace period

**Comment by abasore** [comment]
@slzButton, one way to mitigate the offline grace period from triggering frequently would be to set the policy for Offline Grace Period to be a much larger number than the 720. However, from the video you sent, it looks like the grace period alert comes up with with 0 minutes passes (instead of the expected 720), does that seem correct? When testing for that video is policy still set to 720 minutes and 90 days for grace period wipe?

**Comment by slzButton** [comment]
> @slzButton, one way to mitigate the offline grace period from triggering frequently would be to set the policy for Offline Grace Period to be a much larger number than the 720. However, from the video you sent, it looks like the grace period alert comes up with with 0 minutes passes (instead of the expected 720), does that seem correct? When testing for that video is policy still set to 720 minutes and 90 days for grace period wipe?

我很确定我们设置的离线宽限期是720 minutes and 90 days。之所以会出现视频中的情况，是因为当我们操作过系统时间后，出现了离线宽限期的弹框后，再次把时间设置回来，后续只要出现断网离线，就会复现这个视频中情况，这个bug让我们很痛苦，我们希望能有解决的方案，不管是能让我们的账号设置不使用离线宽限期还是这个bug能被其他方案绕过或修复，都希望能尽快解决一下，
你们可以通过下面的视频操作复现一下试试。
I'm pretty sure we set the offline grace period at 720 minutes and 90 days. The reason for the occurrence of the situation in the video is that when we operate the system time, the pop-up box of offline grace period appears, and then we set the time back. The situation in the video will be repeated as long as the network is disconnected and offline. This bug makes us very painful, and we hope to find a solution. We hope to solve the problem as soon as possible, whether we can make our account Settings not use the offline grace period or this bug can be bypassed or fixed by other solutions.
You can try it out with the video below。

https://user-images.githubusercontent.com/13009437/133871581-52c73dd3-3c4c-4f2b-a649-cef144b062b4.mp4

**Comment by slzButton** [comment]
> @slzButton, one way to mitigate the offline grace period from triggering frequently would be to set the policy for Offline Grace Period to be a much larger number than the 720. However, from the video you sent, it looks like the grace period alert comes up with with 0 minutes passes (instead of the expected 720), does that seem correct? When testing for that video is policy still set to 720 minutes and 90 days for grace period wipe?

**Comment by slzButton** [comment]
> @slzButton, one way to mitigate the offline grace period from triggering frequently would be to set the policy for Offline Grace Period to be a much larger number than the 720. However, from the video you sent, it looks like the grace period alert comes up with with 0 minutes passes (instead of the expected 720), does that seem correct? When testing for that video is policy still set to 720 minutes and 90 days for grace period wipe?

希望能尽快给一个回复，这个问题真的很影响我们的使用

I hope you can give me a reply as soon as possible. This problem really affects our use

**Comment by abasore** [comment]
@slzButton, thanks for providing the videos. The grace period cannot be removed, but the time be set to higher values. I have two questions (below) for you that will hopefully get to the root of the issue.

The expected offline wipe flow is this: (1) disable wifi (2) set the time ahead (3) Navigate back to the app (4) You should then see an "Reconnect to your Organization" alert. You will see either "Sign In" or "Remove Account." For the offline wipe flow you have to sign back in the app (while online!) into the app in order to use it. Following steps 1-3 above, can you confirm if you are able to sign back in after the offline grace period expires as expected while online? 

In the video I see that the you are (1) Signed into the app (2) Set the time ahead by a few months (3) Disabled wifi (4) Opened the app to the "Offline Grace Period Expired" alert (5) Resetting the time back to automatic (6) Opening the app again with wifi enabled. (7) Turn off wifi and are not able to access the app. This appears to be the expected behavior. Can you confirm what behavior you expect? When offline grace period expires, the user must sign back in so they can access company resources. Even if you reset the time to automatic (step 5), the user still has to sign back in online.

**Comment by slzButton** [comment]
> @slzButton, thanks for providing the videos. The grace period cannot be removed, but the time can be set to higher values. I have two questions (below) for you that will hopefully get to the root of the issue.
> 
> The expected offline wipe flow is this: (1) disable wifi (2) set the time ahead (3) Navigate back to the app (4) You should then see an "Reconnect to your Organization" alert. You will see either "Sign In" or "Remove Account." For the offline wipe flow you have to sign back in the app (while online!) into the app in order to use it. Following steps 1-3 above, can you confirm if you are able to sign back in after the offline grace period expires as expected while online?
> 
> In the video I see that the you are (1) Signed into the app (2) Set the time ahead by a few months (3) Disabled wifi (4) Opened the app to the "Offline Grace Period Expired" alert (5) Resetting the time back to automatic (6) Opening the app again with wifi enabled. (7) Turn off wifi and are not able to access the app. This appears to be the expected behavior. Can you confirm what behavior you expect? When offline grace period expires, the user must sign back in so they can access company resources. Even if you reset the time to automatic (step 5), the user still has to sign back in online.

对于第一个问题，我们可以让用户在离线宽限期到期后重新登陆
但是第二个问题，我们疑惑的是，当用户在离线宽限期到期后，再次联网重新进行网络请求之后关闭网络，不是应该进入一个新的离线宽限期循环么？就是说再次离线超过720 分钟和 90 天之后才会提示用户离线宽限期过期，而不是立刻提示用户，因为这样我们的离线宽限期就相当于只保证用户在离线超过宽限期一次后，以后都无法再次离线，这很不符合我们app在离线状态也能打开的原则，我们期望的是，当用户离线宽限期到期后再次联网会重新刷新他的离线宽限期，而不是永远只能离线一次。

For the first problem, we can let the user log in again after the offline grace period expires
But the second question we are wondering is, when the user after the offline grace period expires, and then goes back to the Internet to make the network request again, shouldn't it enter a new offline grace period cycle? Is offline for more than 720 minutes and again after 90 days will prompt the user for offline grace period expires, rather than immediately prompt the user, because it's our offline grace period is equivalent to only guarantee after the grace period once the user in the offline, later can't again offline, which is not in accordance with our app in the offline state can also open the principle, we expect is, When the offline grace period expires, users will be able to renew their offline grace period by connecting to the Internet again, instead of being offline only once.

**Comment by slzButton** [comment]
@abasore Just checking if you are still seeing this issue?

**Comment by slzButton** [comment]
如果我想让我的离线宽限期重置，应该如何操作呢？我调用了这个方法感觉并没有生效。能提示一下我应该怎么解决这个问题么？
What if I want to reset my offline grace period? I called this method and it doesn't feel like it's working. Can you give me some advice on how to solve this problem?

`IntuneMAMEnrollmentManager.instance().deRegisterAndUnenrollAccount((account ?? (IntuneMAMEnrollmentManager.instance().enrolledAccount()) ?? ""), withWipe: true)`

**Comment by abasore** [comment]
> > @slzButton, Yes, once the block grace period expires, the app will not be able to access company data until it goes back online. Once the wipe grace period expires, the app will not be able to access protected data until it goes back online and the user reauthenticates. In order to use the app, the user will need to connect to the internet. In the video, it seem that the offline grace period expires and that is the expected behavior. (Just as a note, that when testing with changing the device time it is often best to reset the device between testing. Setting the device time forward and backwards can result in unexpected behavior)
> > I will not be able to call but you could file a support ticket for this to get more assistance. For a support ticket you can go to Endpoint Manager (https://endpoint.microsoft.com) to file a support ticket ("Troubleshooting + support" -> "Help and support" -> "Intune").
> 
> @abasore 十分感谢您的回复，我大概对我视频中的设备出现的情况有所了解了，还有一个小问题，如果我想让我视频中的设备重新恢复离线宽限期，就是使这个设备重置它的离线宽限期，我应该怎么操作呢？
> 
> `IntuneMAMEnrollmentManager.instance().deRegisterAndUnenrollAccount((account ?? (IntuneMAMEnrollmentManager.instance().enrolledAccount()) ?? ""), withWipe: true)`
> 
> 我调用了这个方法感觉并没有生效，我是希望这个设备能够恢复到原来有离线宽限期的状态。


@slzButton, In order to reset the offline grace period the user needs to take the device back online and re-authenticate. As for the EnrollmentManager method, ".deRegisterAndUnenrollAccount" can be used to remove the provided account from the list of registered accounts. In order for the user's to use the app with policies, they would then have to be re-enrolled with "registerAndEnrollAccount" or "loginAndEnrollAccount." But again, they will require the user to take device back online and to re-authenticate.

**Comment by abasore** [comment]
@slzButton, it may be easier for us to provide support for this issue through a support ticket. By filing a support ticket, we could look at logs and the details of your app's implementation. For a support ticket you can go to Endpoint Manager (https://endpoint.microsoft.com) to file a support ticket ("Troubleshooting + support" -> "Help and support" -> "Intune"). Let me know if you are having any trouble filing the ticket.

**Comment by slzButton** [comment]
> > > @slzButton, Yes, once the block grace period expires, the app will not be able to access company data until it goes back online. Once the wipe grace period expires, the app will not be able to access protected data until it goes back online and the user reauthenticates. In order to use the app, the user will need to connect to the internet. In the video, it seem that the offline grace period expires and that is the expected behavior. (Just as a note, that when testing with changing the device time it is often best to reset the device between testing. Setting the device time forward and backwards can result in unexpected behavior)
> > > I will not be able to call but you could file a support ticket for this to get more assistance. For a support ticket you can go to Endpoint Manager (https://endpoint.microsoft.com) to file a support ticket ("Troubleshooting + support" -> "Help and support" -> "Intune").
> > 
> > 
> > @abasore 十分感谢您的回复，我大概对我视频中的设备出现的情况有所了解了，还有一个小问题，如果我想让我视频中的设备重新恢复离线宽限期，就是使这个设备重置它的离线宽限期，我应该怎么操作呢？
> > `IntuneMAMEnrollmentManager.instance().deRegisterAndUnenrollAccount((account ?? (IntuneMAMEnrollmentManager.instance().enrolledAccount()) ?? ""), withWipe: true)`
> > 我调用了这个方法感觉并没有生效，我是希望这个设备能够恢复到原来有离线宽限期的状态。
> 
> @slzButton, In order to reset the offline grace period the user needs to take the device back online and re-authenticate. As for the EnrollmentManager method, ".deRegisterAndUnenrollAccount" can be used to remove the provided account from the list of registered accounts. In order for the user's to use the app with policies, they would then have to be re-enrolled with "registerAndEnrollAccount" or "loginAndEnrollAccount." But again, they will require the user to take device back online and to re-authenticate.


我们尝试了在联网状态下使用.deRegisterAndUnenrollAccount方法，之后再次使用.loginAndEnrollAccount让用户再次绑定，可是当用户离线的时候还是出现离线宽限期到期的提示，有没有别的方案能让离线宽限期重置的呢？

We tried using in network condition. DeRegisterAndUnenrollAccount method, again after use. LoginAndEnrollAccount allows users to bind again, but when the user is offline or appears offline grace period expires, Is there another way to reset the offline grace period?

**Comment by abasore** [comment]
@slzButton No, once the **block** offline grace period has expired the user must connect to network access in organ to continue app access. Similarly once the **wipe** offline grace period expires, users must connect to a network and reauthenticate in order to continue app access.

**Comment by abasore** [comment]
Closing this issue for now. Feel free to reopen or open a new issue if problem persists.
